File: MonteCarlo/RunMC.m
% Simple Monte Carlo Campaign

% ... Initialization and Setup
      addpath('../Utilities');
      more off  % So octave won't buffer screen output
      %rand('seed',1);  % For reproducible runs, set seed
      BasePath = './Baseline';
      Nrun = 20;
      MaxCore = 6;

      Nbatch = ceil(Nrun/MaxCore);

      % Allocate output variables
      SettleTime = zeros(Nrun,1);

% ... Campaign Loop
      for Ibatch = 1:Nbatch,
         Ncore = min(MaxCore,Nrun-(Ibatch-1)*MaxCore);
         BatchPath = '';
         for Icore = 1:Ncore,
            Irun = (Ibatch-1)*MaxCore + Icore;

            % Copy Base folder to Run folder
            RunPath = sprintf('Run%03d',Irun);
            system(['rm -rf ',RunPath]);
            system(['cp -r ',BasePath,' ',RunPath]);
            BatchPath = [BatchPath,' ',RunPath];
            
            % Make sure graphics front end is FALSE.
            s = sprintf('FALSE                           !  Graphics Front End?\n');
            OverwriteLineInFile([RunPath,'/Inp_Sim.txt'],6,s);

            % Modify Input Files as desired
            EulAng = 20*(2*rand(3,1)-1);
            s = sprintf(...
               '%8.2f  %8.2f   %8.2f    213      ! Angles (deg) & Euler Sequence\n',...
               EulAng(1),EulAng(2),EulAng(3));
            OverwriteLineInFile([RunPath,'/SC_Simple.txt'],16,s);

         end

         % Perform Runs
         fprintf('Starting Runs %d - %d\n',(Ibatch-1)*MaxCore + 1, (Ibatch-1)*MaxCore+Ncore);
         system(['../Utilities/BatchRun.sh ',BatchPath]);

         % Process Batch Results
         for Icore = 1:Ncore,
            Irun = (Ibatch-1)*MaxCore + Icore;
            RunPath = sprintf('Run%03d',Irun);

            % Process Run in Matlab/Octave
            LoadString = [RunPath,'/time.42'];
            load(LoadString);
            LoadString = [RunPath,'/wbn.42'];
            load(LoadString);

            N = length(time);
            for k=1:N-1,
               magwbn = norm(wbn(k,:));
               if (magwbn > 1.0E-6*180/pi),
                  SettleTime(Irun) = time(k+1);
               end
            end

            % Save only selected *.42 files
            SaveList = {'time','wbn'};
            Nsave = length(SaveList);
            for Isave = 1:Nsave,
               system(['mv ',RunPath,'/',SaveList{Isave},'.42 ',RunPath,'/',SaveList{Isave},'.tmp']);
            end
            system(['rm ',RunPath,'/*.42']);
            for Isave = 1:Nsave,
               system(['mv ',RunPath,'/',SaveList{Isave},'.tmp ',RunPath,'/',SaveList{Isave},'.42']);
            end

            % Delete unaltered *.txt files
            system(['rm ',RunPath,'/Flex_Simple.txt']);
            system(['rm ',RunPath,'/Inp_Cmd.txt']);
            system(['rm ',RunPath,'/Inp_FOV.txt']);
            system(['rm ',RunPath,'/Inp_Graphics.txt']);
            system(['rm ',RunPath,'/Inp_IPC.txt']);
            system(['rm ',RunPath,'/Inp_NOS3.txt']);
            system(['rm ',RunPath,'/Inp_Region.txt']);
            %system(['rm ',RunPath,'/Inp_Sim.txt']);
            system(['rm ',RunPath,'/Inp_TDRS.txt']);
            system(['rm ',RunPath,'/Orb_LEO.txt']);
            %system(['rm ',RunPath,'/SC_Simple.txt']);

         end

      end

% ... Post-processing
      figure(1)
      hist(SettleTime)
      title('Distribution of Settle Times')
      xlabel('Settle Time, sec')
      ylabel('Occurrences')
      save -ascii SettleTime.txt SettleTime;

% ... Cleanup
      more on
      rmpath('../Utilities');

File: MonteCarlo/RunMC.jl
using DelimitedFiles;
using Dates;
using Plots;
gr();

################################################################################
function ReplaceLineInFile(FileName,LineNum,text)

      (tmppath,tmpio) = mktemp();
      infile = open(FileName,"r");
      for i = 1:LineNum-1
         line = readline(infile,keep=true);
         write(tmpio,line);
      end
      line = readline(infile,keep=true);
      write(tmpio,text*"\n");
      while !eof(infile)
         line = readline(infile,keep=true);
         write(tmpio,line);
      end
      close(infile);
      close(tmpio);
      mv(tmppath,FileName,force=true);
            
end;
################################################################################
# Simple Monte Carlo Campaign 
function RunMC()

# ... Initialization and Setup
      BasePath = "./Baseline";
      Nrun = 20;
      MaxCore = 12;
      
      # Allocate output variables
      SettleTime = zeros(Nrun);

# ... Campaign Loop
      Ndigits = Int(floor(log10(Nrun)))+1;
      Nbatch = Int(ceil(Nrun/MaxCore));
      StartTime = now();
      # Make sure graphics front end is FALSE
      s = "FALSE                           !  Graphics Front End?";
      ReplaceLineInFile(BasePath*"/Inp_Sim.txt",7,s);
            
      # Prepare batch inputs
      for Ibatch = 1:Nbatch
         if Nrun-(Ibatch-1)*MaxCore < MaxCore
            Ncore = Nrun-(Ibatch-1)*MaxCore
         else
            Ncore = MaxCore
         end
         BatchPath = "";
         RunPath = Vector{String}(undef,Ncore);
         RunProc = Vector{Any}(undef,Ncore);
         for Icore = 1:Ncore
            Irun = Int((Ibatch-1)*MaxCore+Icore);
            
            # Copy Base folder to Run folder
            RunPath[Icore] = "Run"*lpad(Irun,Ndigits,"0");
            cp(BasePath,RunPath[Icore],force=true);
            BatchPath = BatchPath*" "*RunPath[Icore];
            
            # Modify input files as desired
            EulAng = 20*(2*rand(3) .- 1);
            EulAng = round.(EulAng,digits=2);
            s = "$(EulAng[1]) $(EulAng[2]) $(EulAng[3])    213      ! Angles (deg) & Euler Sequence";
            ReplaceLineInFile(RunPath[Icore]*"/SC_Simple.txt",16,s);
            
         end

         # Perform runs
         FirstRunOfBatch = (Ibatch-1)*MaxCore+1;
         LastRunOfBatch = (Ibatch-1)*MaxCore+Ncore;
         println("Starting Runs $(FirstRunOfBatch) - $(LastRunOfBatch)");
         cd("..")
         t1 = now();
         #for Icore = 1:Ncore-1
         #   Path = "MonteCarlo/$(RunPath[Icore])"
         #   run(pipeline(`./42 $Path`,stdout="$Path/stdout.txt",stderr="$Path/stderr.txt"),wait=false);
         #end
         #Path = "MonteCarlo/$(RunPath[Ncore])"
         #run(pipeline(`./42 $Path`,stdout="$Path/stdout.txt",stderr="$Path/stderr.txt"),wait=true);
         for Icore = 1:Ncore
            Path = "MonteCarlo/$(RunPath[Icore])"
            RunProc[Icore] = run(pipeline(`./42 $Path`,stdout="$Path/stdout.txt",stderr="$Path/stderr.txt"),wait=false);
         end
         for Icore = 1:Ncore
            wait(RunProc[Icore]);
         end
         t2 = now();
         dt = Dates.value(t2-t1)/1000.0;
         println("All runs complete in $dt sec");
         cd("MonteCarlo");
      
         # Process batch results
         for Icore = 1:Ncore
            Irun = (Ibatch-1)*MaxCore + Icore;
            cd(RunPath[Icore]);
            time = readdlm("time.42");
            wbn = readdlm("wbn.42");

            N = length(time);
            for k=1:N-1,
               magwbn = sqrt(wbn[k,1]^2+wbn[k,2]^2+wbn[k,3]^2);
               if (magwbn > 1.0E-4*180/pi)
                  SettleTime[Irun] = time[k+1];
               end
            end
            
            # Save only selected .42 files
            SaveList = ["time","wbn"];
            dir = readdir();
            for i = 1:length(dir)
               if occursin(".42",dir[i]) && !any(occursin.(SaveList,dir[i]))
                  rm(dir[i]);
               end
            end
            
            # Delete unaltered *.txt files
            rm("Flex_Simple.txt");
            rm("Inp_Cmd.txt");
            rm("Inp_FOV.txt");
            rm("Inp_Graphics.txt");
            rm("Inp_IPC.txt");
            rm("Inp_NOS3.txt");
            rm("Inp_Region.txt");
            rm("Inp_Sim.txt");
            rm("Inp_TDRS.txt");
            rm("Nodes_Simple.txt");
            rm("Orb_LEO.txt");
            #rm("SC_Simple.txt");
            rm("Shaker_Simple.txt");
            rm("TRV.txt");
            rm("Whl_Simple.txt");
            
            cd("..")
         end
         
      end
            
# ... Post-processing
      open("SettleTime.txt","w") do io
         writedlm(io,SettleTime,' ');
      end
      
      EndTime = now();
      dt = Dates.value(EndTime-StartTime)/1000.0;
      println("Campaign Duration: $dt sec");

end
################################################################################
if abspath(PROGRAM_FILE) == @__FILE__
      RunMC();
end;
